File: core/src/com/xenoage/zong/core/music/util/BeatE.java
package com.xenoage.zong.core.music.util;

import lombok.Data;
import lombok.NonNull;

import com.xenoage.utils.math.Fraction;


/**
 * Element at a beat.
 * 
 * This is a wrapper class to combine any object with
 * the beat (represented as a {@link Fraction}) it belongs to.
 * 
 * @author Andreas Wenger
 */
@Data public final class BeatE<T> {

	/** The element at the beat */
	@NonNull public final T element;
	/** The beat where the element can be found */
	@NonNull public final Fraction beat;
	
	
	public static <T> BeatE<T> beatE(T element, Fraction beat) {
		return new BeatE<T>(element, beat);
	}


	/**
	 * Returns the latest of the given elements. If none are given or
	 * no items with beats are given, null is returned.
	 */
	@SafeVarargs public static <T2> BeatE<T2> selectLatest(BeatE<T2>... elements) {
		BeatE<T2> ret = null;
		for (BeatE<T2> element : elements) {
			if (ret == null || (element != null && element.beat.compareTo(ret.beat) > 0))
				ret = element;
		}
		return ret;
	}


	@Override public String toString() {
		return element + " at beat " + beat;
	}


}

File: core/src/com/xenoage/zong/core/music/util/BeatE.kt
package com.xenoage.zong.core.music.util

import com.xenoage.zong.core.position.Beat

/**
 * Element at a beat.
 *
 * This is a wrapper class to combine any object with
 * the beat it belongs to.
 */
data class BeatE<T>(
	/** The element at the beat  */
	val element: T,
	/** The beat where the element can be found  */
	val beat: Beat
) {

	override fun toString() = "$element at beat $beat"

	companion object {
		/**
		 * Returns the latest of the given elements. If none are given or
		 * no items with beats are given, null is returned.
		 */
		fun <T2> selectLatest(vararg elements: BeatE<T2>): BeatE<T2>? {
			var ret: BeatE<T2>? = null
			for (element in elements) {
				if (ret == null || element.beat.compareTo(ret.beat) > 0)
					ret = element
			}
			return ret
		}
	}

}
